### PR TITLE
basic mvp flow

### PR DESCRIPTION
**Pull request by Matias-Villalobo**
As you can see, mvp model has been implemented. And now when you press the numbers, they are shown in the calculator screen.

https://user-images.githubusercontent.com/73836200/105217365-83a43000-5b32-11eb-8794-e000d38c0541.mp4

### CODE REVIEW

File: app/src/main/java/com/example/mykotlincalculator21/mvp/model/CalculatorModel.kt [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+import com.example.mykotlincalculator21.mvp.contract.CalculatorContract
+import com.example.mykotlincalculator21.utils.EMPTY_STRING
+
+class CalculatorModel : CalculatorContract.CalculatorModelContract {
+    override var firstOperand: String = EMPTY_STRING
+
+    override fun saveNumber(number: String) {
+        firstOperand = number
+    }
+
```
Comment (lucianoeli) [inline app/src/main/java/com/example/mykotlincalculator21/mvp/model/CalculatorModel.kt:12]: delete this blank line plis

File: app/src/main/java/com/example/mykotlincalculator21/mvp/presenter/CalculatorPresenter.kt [hunk 2]
```diff
@@ -0,0 +1,19 @@
+package com.example.mykotlincalculator21.mvp.presenter
+
+import com.example.mykotlincalculator21.mvp.contract.CalculatorContract
+
+class CalculatorPresenter(model: CalculatorContract.CalculatorModelContract, view: CalculatorContract.CalculatorViewContract): CalculatorContract.CalculatorPresenterContract {
+
+    private val model: CalculatorContract.CalculatorModelContract
+    private val view: CalculatorContract.CalculatorViewContract
+
+    init{
```
Comment (lucianoeli) [inline app/src/main/java/com/example/mykotlincalculator21/mvp/presenter/CalculatorPresenter.kt]: this should be init { (with the space before bracket)

remember Ctrl + Alt + L and Ctrl Alt + O

File: app/src/main/java/com/example/mykotlincalculator21/mvp/view/ActivityView.kt [hunk 3]
```diff
@@ -0,0 +1,26 @@
+package com.example.mykotlincalculator21.mvp.view
+
+
+import android.app.Activity
+
```
Comment (lucianoeli) [inline app/src/main/java/com/example/mykotlincalculator21/mvp/view/ActivityView.kt]: delete unneccessary empty lines

File: app/src/main/java/com/example/mykotlincalculator21/mvp/model/CalculatorModel.kt [hunk 4]
```diff
@@ -0,0 +1,13 @@
+package com.example.mykotlincalculator21.mvp.model
+
+import com.example.mykotlincalculator21.mvp.contract.CalculatorContract
+import com.example.mykotlincalculator21.utils.EMPTY_STRING
+
+class CalculatorModel : CalculatorContract.CalculatorModelContract {
+    override var firstOperand: String = EMPTY_STRING
```
Comment (lucianoeli) [inline app/src/main/java/com/example/mykotlincalculator21/mvp/model/CalculatorModel.kt]: dont put properties on the contract. they shoud be private

Comment (SartiMau) [reply app/src/main/java/com/example/mykotlincalculator21/mvp/model/CalculatorModel.kt]: if he will use the variable for the native get and set I think it is okay, but as he is setting this variable with the method **saveNumber** the variable shouldn't be on the contract

File: app/src/main/java/com/example/mykotlincalculator21/mvp/presenter/CalculatorPresenter.kt [hunk 5]
```diff
@@ -0,0 +1,19 @@
+package com.example.mykotlincalculator21.mvp.presenter
+
+import com.example.mykotlincalculator21.mvp.contract.CalculatorContract
+
+class CalculatorPresenter(model: CalculatorContract.CalculatorModelContract, view: CalculatorContract.CalculatorViewContract): CalculatorContract.CalculatorPresenterContract {
+
+    private val model: CalculatorContract.CalculatorModelContract
```
Comment (SartiMau) [inline app/src/main/java/com/example/mykotlincalculator21/mvp/presenter/CalculatorPresenter.kt]: you can remove these variables and add private val before each one in the class constructor

class CalculatorPresenter(private val model: CalculatorContract.CalculatorModelContract, private val view: CalculatorContract.CalculatorViewContract)

File: app/src/main/java/com/example/mykotlincalculator21/mvp/view/ActivityView.kt [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+import android.content.Context
+import androidx.annotation.Nullable
+import java.lang.ref.WeakReference
+
+open class ActivityView(activity:Activity) {
+    private val activityRef:WeakReference<Activity>
+    val activity: Activity?
+        @Nullable
```
Comment (SartiMau) [inline app/src/main/java/com/example/mykotlincalculator21/mvp/view/ActivityView.kt]: the annotation is not necessary here

File: app/src/main/java/com/example/mykotlincalculator21/mvp/view/ActivityView.kt [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+        get() {
+            return activityRef.get()
+        }
+    val context: Activity?
+        @Nullable
```
Comment (SartiMau) [inline app/src/main/java/com/example/mykotlincalculator21/mvp/view/ActivityView.kt]: same here

File: app/src/main/java/com/example/mykotlincalculator21/mvp/view/CalculatorView.kt [hunk 8]
```diff
@@ -0,0 +1,18 @@
+package com.example.mykotlincalculator21.mvp.view
+
+import com.example.mykotlincalculator21.MainActivity
+import com.example.mykotlincalculator21.databinding.ActivityMainBinding
+import com.example.mykotlincalculator21.mvp.contract.CalculatorContract
+
+class CalculatorView(mainActivity: MainActivity, binding: ActivityMainBinding): ActivityView(mainActivity), CalculatorContract.CalculatorViewContract {
```
Comment (SartiMau) [inline app/src/main/java/com/example/mykotlincalculator21/mvp/view/CalculatorView.kt]: same here.. you can add private val in the constructor and delete the init statement and the variable

**Review by SartiMau — CHANGES_REQUESTED** [review]

**Review by SartiMau — APPROVED** [review]

File: app/src/main/java/com/example/mykotlincalculator21/mvp/model/CalculatorModel.kt (hunk 1 above)
Comment (Matias-Villalobo) [reply app/src/main/java/com/example/mykotlincalculator21/mvp/model/CalculatorModel.kt:12]: Done. Thx

File: app/src/main/java/com/example/mykotlincalculator21/mvp/presenter/CalculatorPresenter.kt (hunk 2 above)
Comment (Matias-Villalobo) [reply app/src/main/java/com/example/mykotlincalculator21/mvp/presenter/CalculatorPresenter.kt]: Done. Thx

File: app/src/main/java/com/example/mykotlincalculator21/mvp/view/ActivityView.kt (hunk 3 above)
Comment (Matias-Villalobo) [reply app/src/main/java/com/example/mykotlincalculator21/mvp/view/ActivityView.kt]: Done. Thx

File: app/src/main/java/com/example/mykotlincalculator21/mvp/model/CalculatorModel.kt (hunk 4 above)
Comment (Matias-Villalobo) [reply app/src/main/java/com/example/mykotlincalculator21/mvp/model/CalculatorModel.kt]: Done. Thx guys, now its fixed

File: app/src/main/java/com/example/mykotlincalculator21/mvp/presenter/CalculatorPresenter.kt (hunk 5 above)
Comment (Matias-Villalobo) [reply app/src/main/java/com/example/mykotlincalculator21/mvp/presenter/CalculatorPresenter.kt]: Done. Thx

File: app/src/main/java/com/example/mykotlincalculator21/mvp/view/ActivityView.kt (hunk 6 above)
Comment (Matias-Villalobo) [reply app/src/main/java/com/example/mykotlincalculator21/mvp/view/ActivityView.kt]: Done. Thx

File: app/src/main/java/com/example/mykotlincalculator21/mvp/view/ActivityView.kt (hunk 7 above)
Comment (Matias-Villalobo) [reply app/src/main/java/com/example/mykotlincalculator21/mvp/view/ActivityView.kt]: Done. Thx

File: app/src/main/java/com/example/mykotlincalculator21/mvp/view/CalculatorView.kt (hunk 8 above)
Comment (Matias-Villalobo) [reply app/src/main/java/com/example/mykotlincalculator21/mvp/view/CalculatorView.kt]: Done. Thx

**Review by BQuinteros — APPROVED** [review]

File: app/src/main/java/com/example/mykotlincalculator21/mvp/presenter/CalculatorPresenter.kt [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+import com.example.mykotlincalculator21.mvp.contract.CalculatorContract
+
+class CalculatorPresenter(
+    private val model: CalculatorContract.CalculatorModelContract,
+    private val view: CalculatorContract.CalculatorViewContract
+) : CalculatorContract.CalculatorPresenterContract {
+
+    override fun numberPressed(number: String) {
+        model.saveNumber(number)
+        view.drawNumber(model.getResult()!!)
```
Comment (nicolbidone) [inline app/src/main/java/com/example/mykotlincalculator21/mvp/presenter/CalculatorPresenter.kt]: please avoid using !!

File: app/src/main/java/com/example/mykotlincalculator21/utils/StringUtils.kt [hunk 10]
```diff
@@ -0,0 +1,5 @@
+package com.example.mykotlincalculator21.utils
+
+const val EMPTY_STRING = ""
+
+
```
Comment (BQuinteros) [inline app/src/main/java/com/example/mykotlincalculator21/utils/StringUtils.kt]: pls remove extra empty line

**Review by BQuinteros — APPROVED** [review]

**Review by lucianoeli — APPROVED** [review]

**Review by SartiMau — APPROVED** [review]

**Review by nicolbidone — APPROVED** [review]

**Pull request merged**